### ISSUE TITLE
Dataset Functionality

### ISSUE THREAD

**Issue by vincentqb** [pytorch/pytorch#24915]
## 🚀 Feature

The Domain APIs all have separate interface to acquire data. We are discussing having:
- [ ] all datasets moved from Domain APIs to PyTorch
- [ ] a core download feature, within PyTorch
- [ ] a core raw data loading (e.g. ByteTensor iterator), within PyTorch
- [ ] domain-specific pre-processing in each Domain APIs
- [ ] tutorials on how to build data pipeline

## Motivation

* The Domain APIs currently each have their own non-standard dataset structure that may also download the data. This duplicate efforts and adds complexity to the user.
* A common bottleneck when generating datasets is reading the data. We want to offer an interface that enables reading the data and running initial preprocessing while maximizing available computing resources utilization.
* We want to leverage specialize libraries such as NVIDIA DALI.

## Alternatives

Status quo: fragmented effort.
* torchvision @fmassa
* torchtext @zhangguanheng66 
* torchaudio @vincentqb

## Additional Information

@pbelevich -- for dataset abstraction
@cpuhrsch -- for NestedTensor

#12672

**Comment by vadimkantorov** [comment]
An additional wish: unify transforms interface. One idea: make them regular autograd functions / modules. This enables writing them more shared between domains (imagine various forms of cutout / warps), moving them from cpu to gpu, performing them at other parts of the model, etc

**Comment by soumith** [comment]
all datasets are not moving to core pytorch. If you want, create a `torchdata` for it and make domain APIs depend on `torchdata`.

**Comment by zhangguanheng66** [comment]
> all datasets are not moving to core pytorch. If you want, create a `torchdata` for it and make domain APIs depend on `torchdata`.

One of the plans is to keep the raw data links and utils in pytorch/data and have the preprocessing func and associated ops in each domain.

**Comment by SsnL** [comment]
I'm wonder what functionalities are planned for `torchdata`. Most of the proposed features sound already doable nowadays. I think additional explanation on what the exact needs are and why the current code can't do the job will be helpful for discussion.

For example,

>  a core download feature, within PyTorch

sounds similar to `torch.hub`'s download utility function

> a core raw data loading (e.g. ByteTensor iterator), within PyTorch

 sounds doable just with the current data loading infra.

**Comment by vadimkantorov** [comment]
@soumith Any thoughts on my suggestion about `transforms` interface? (i.e. making them normal pytorch functions / modules to support better reuse)

**Comment by fmassa** [comment]
@vadimkantorov 
> Any thoughts on my suggestion about transforms interface? (i.e. making them normal pytorch functions / modules to support better reuse)

Yes, I agree, and torchvision is going to be doing this, see https://github.com/pytorch/vision/issues/1375

**Comment by vincentqb** [comment]
Thoughts from offline chat with @taylorgordon20.

How to help in debugging in a natural python way?
* Could use maps (e.g. `Map(read, get_line)`) instead of generators.
* Do we want static type analysis?
* Enforce structure between generators?
* `tensorflow_datasets/image/coco.py` has "features" for typing
* [facebook pyre](https://pyre-check.org/) may support templating and "typed dictionary" check in functions? 

Wishes:
* Batched
* Lazy better than Eager for data loading
* Async preferred over Sync
* Prefer imperative over declarative
* Serializable can be done in a few ways: torchscript, etc
* Shuffle, Sample
* Save, cache transformed data
